File: Leo Kim/2023-01-30-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:

        ## key point of this algorithm is:
        ## nums[middle] >= nums[left]: serach from right else from left

        l = 0
        r = len(nums) - 1

        while l <= r:
            mid = (l + r) // 2
            if nums[mid] < nums[r]:
                r = mid
            else:
                l = mid + 1

        return nums[r]

        # TC: logN usual bintree tc

        # return min(sorted(nums)) This is faster... SO FUNNY HAHA

File: MinjiJu/2023-01-30-Find-Minimum-in-Rotated-Sorted-Array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int left=0, right=nums.size()-1;
        
        while(left<right){
            auto mid = left+(right-left)/2;
            
            if(nums[mid]<nums[right]) right=mid;
            else if(nums[mid]>nums[right]) left=mid+1;
        }
        
        return nums[left];
    }
};